### DISCUSSION TITLE
Generator functions

### DISCUSSION THREAD

**Discussion by lilactown** [clavascript/clavascript#186] [Ideas]
## Proposal

```clojure
(defn ^:gen my-generator []
  (yield 1)
  (yield 2)
  (yield 3))
```

Potential alternatives: `defn*`, `defn ^:generator`

## Rationale

Generators are a fundamental construct in JS that enables novel evaluation semantics. Fundamentally they are lazy, stateful state machines written with a single keyword `yield`. The following use cases use generators as a building block

* Building infinite sequences (see `LazyIterable` in this project)
* [js-csp](https://github.com/js-csp/js-csp), a port of `core.async` to JS
* [crank.js](https://crank.js.org), a React alternative that uses generators for stateful components

### Building lazy iterables

In Clojure, there is `lazy-seq` for building custom lazy sequences. We could provide a simple alternative: a `lazy` function that accepts a generator functions

```clojure
(defn ^:gen -foo []
  (yield 1)
  (yield 2)
  (yield 3))

(def foo (lazy -foo))

(vec (map inc (foo)))
;; => [1 2 3]

(vec (filter even? (foo)))
;; => [2]
```

### Async generators

Whatever syntax we pick should interop with async functions.

```clojure
(defn ^:async ^:gen fetch-pages []
  (loop [page 0]
    (let [data (await (fetch-page page)]
      (yield data)
      (recur (inc page)))))
```

**Comment by borkdude** [comment]
@lilactown Looks good to me, but I propose we write `(js/yield ...)` and `(js/await ...)`, since those symbols would be unresolved with clj-kondo otherwise?

**Reply by borkdude** [reply]
This is implemented in the compiler by making `js/await` a special form.

**Reply by lilactown** [reply]
I've been wondering why we necessitate the `js` prefix. Is it because of clj-kondo?

**Reply by borkdude** [reply]
Yes

**Reply by borkdude** [reply]
<img width="398" alt="Screenshot 2022-08-30 at 17 32 37" src="https://user-images.githubusercontent.com/284934/187479003-7edce5f0-9aae-430a-a489-2bf6f0135a7f.png">

**Reply by lilactown** [reply]
We can add custom support for Clava :)

**Reply by borkdude** [reply]
Maybe eventually, but I'd rather delay that until we have fleshed out things more.
Another option could be this:

```
(:require [clavascript.core :refer [await yield]])
```
so teach kondo where they come from.

**Reply by borkdude** [reply]
(which would just be a no-op for clavascript)

**Reply by borkdude** [reply]
If we would have a specific extension for clavascript, `.clvs`, we could teach clj-kondo eventually, or so.
